src/pagecache/mod.rs

```diff
@@ -2076,4 +2075,37 @@ impl PageCache {
 
         Ok(())
     }
+
+    /// A snapshot is to recover the pageTable at a point in time.
+    ///
+    /// This is called fuzzy snapshot because while
+    /// we are taking a snapshot, the ongoing inserts
+    /// into the database will keep bumping up the Lsn.
+    /// Therefore, the `stable_lsn` gives us the state
+    /// of the world, when the snapshot was taken.
+    #[allow(unused)]
+    fn take_fuzzy_snapshot(self) -> Snapshot {
+        let stable_lsn_now: Lsn = self.log.stable_offset();
+
+        // This is how we determine the number of the pages we will snapshot.
+        let pid_bound = self.next_pid_to_allocate.load(Acquire);
+
+        let pid_bound_usize = assert_usize(pid_bound);
+
+        let mut page_states =
+            Vec::<PageState>::with_capacity(pid_bound_usize);
+        let guard = pin();
+        for pid in 0..pid_bound {
+            if let Some(pg_view) = self.inner.get(pid, &guard) {
+                let page_state = pg_view.to_page_state();
+                page_states.push(page_state);
+            }
```

Comment:
this looks great overall! there's just two subtle race conditions I realized though while reading this:

1. some other thread is going to allocate a new page, so they bump the `next_pid_to_allocate`. they have not yet installed anything into the `PageTable` though, so if anyone were to try to get the page, it will return `None`.
2. the allocating thread installs an empty pagetable entry before calling cas_page on it to actually install the data. if this thread calls to_page_state on it, it will panic due to array indexing 0 on an empty array. 

I'll fix both of these by looping until both of the above conditions are false, since the allocating thread will fix both of them on its own and we don't have the ability to cooperatively complete their operation without knowing the page they are trying to allocate.

These don't need to be fixed in this PR, I'll do it after merging.

Reply:
Sounds good